docs/repair_based_node_ops.md

```diff
@@ -0,0 +1,101 @@
+# Repair based node operations
+
+Here below is a simple introduction to the node operations Scylla support and
+some of the issues that streaming based node operations have.
+
+- Replace operation
+
+   It is used to replace a dead node. The token ring does not change. Replacing
+   node pulls data from only one of the replicas which might not be the latest
+   copy.
+
+- Rebuild operation
+
+   It is used to get all the data this node owns form other existing nodes. It
+   pulls data from only one of the replicas which might not be the latest copy.
+
+- Removenode operation
+
+   It is used to remove a dead node out of the cluster. Existing nodes pull
+   data from other existing nodes for the new ranges it owns. It pulls from one
+   of the replicas which might not be the latest copy.
+
+- Bootstrap operation
+
+   It is used to add a new node into the cluster. The token ring changes. It
+   does not suffer from the "latest replica” issue. New node pulls data
+   from existing nodes that are losing the token ranges.
+
+   It suffers from failed streaming. We split the ranges in 10 groups and we
+   stream one group at a time. Restream the group if failed, causing
+   unnecessary data transmission on wire.
+
+   Bootstrap is not resumable. With failure after 99.99% of data is streamed,
+   if we restart the node again, we need to stream all the data again even if
+   the node already has 99.99% of the data.
+
+- Decommission operation
+
+   It is used to remove a live node form the cluster. Token ring changes. It
+   does not suffer from the “latest replica” issue. The leaving node pushes data
+   to existing nodes.
+
+   It suffers from resumable issue like bootstrap operation.
+
+To solve all the issues above. We use repair based node operations. The idea
+behind repair based node operations is simple: use repair to synchronize data
+between replicas instead of using streaming.
+
+The benefits:
+
+- Latest copy is guaranteed
+
+- Resumable in nature
+
+- No extra data is streamed on wire
+  E.g., rebuild twice, will not stream the same data twice
+
+- Unified code path for all the node operations
+
+- Free repair operation during bootstrap, replace operation and so on.
+
+Select nodes to synchronize with:
+
+- Replace operation
+  Synchronize with all replica nodes from the local DC
+
+- Rebuild operation
+  Synchronize with all replica nodes from the local DC or from the DC specified by the user
+
+- Removenode operation
+  Synchronize with all replica nodes from local DC
+
+- Bootstrap operation
+  1) If local_dc_replica_nodes = RF, synchronize with 1 node that is losing the
+  range in the local DC
+
```

Comment:
What if that node is down? Or does local_dc_replica_nodes only includes live nodes?

Reply:
In this case, the bootstrap will fail.  If we continue, we would lose the replica on the node losing the range.  Bootstrap can sync with one node instead of only quorum nodes replys on the fact that during bootstrap, the node "transfer" the data to the new node.

Reply:
But we can tranfer from the node that is losing the range OR a quorum of other nodes.

Reply:
When we have enough replica nodes, the best option is to recover the dead node then do the one node transfer bootstrap. When there are not enough replicas nodes, we have no choice but transfer with quorum nodes. We can add an option to make one node transfer case fallback to quorum transfer but I think we should not do it automatically.